File: frontend/src/lib/analytics.ts
import type { FormFieldDef } from "@/types/form";
import type {
  AnalyticsSummary,
  QuestionAnalytics,
  ResponseAnswers,
} from "@/types/response";

export interface EventInput {
  eventType: string;
  step: number | null;
  fieldId: string | null;
  sessionId: string | null;
  createdAt: number;
}

export interface ResponseInput {
  submittedAt: number;
  completionTimeSeconds: number | null;
  answers: ResponseAnswers;
}

interface ComputeAnalyticsInput {
  fields: FormFieldDef[];
  events: EventInput[];
  responses: ResponseInput[];
  rangeDays?: number;
}

export function computeAnalytics({
  fields,
  events,
  responses,
  rangeDays = 14,
}: ComputeAnalyticsInput): AnalyticsSummary {
  const totalViews = events.filter((e) => e.eventType === "view").length;
  const totalStarts = events.filter((e) => e.eventType === "start").length;
  const totalSubmissions = responses.length;

  const completionRate = totalStarts
    ? Math.min(100, Math.round((totalSubmissions / totalStarts) * 100))
    : totalViews
    ? Math.round((totalSubmissions / totalViews) * 100)
    : 0;

  const completionTimes = responses
    .map((r) => r.completionTimeSeconds)
    .filter((v): v is number => typeof v === "number" && v > 0);
  const avgCompletionSeconds = completionTimes.length
    ? Math.round(
        completionTimes.reduce((a, b) => a + b, 0) / completionTimes.length,
      )
    : 0;

  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const responsesByDay: { date: string; count: number }[] = [];
  for (let i = rangeDays - 1; i >= 0; i--) {
    const day = new Date(today);
    day.setDate(today.getDate() - i);
    const key = day.toISOString().slice(0, 10);
    const count = responses.filter((r) => {
      const d = new Date(r.submittedAt);
      d.setHours(0, 0, 0, 0);
      return d.toISOString().slice(0, 10) === key;
    }).length;
    responsesByDay.push({ date: key, count });
  }

  const stepNumbers = Array.from(
    new Set(fields.map((f) => f.step).filter((s) => s > 0)),
  ).sort((a, b) => a - b);
  const dropoffByStep = stepNumbers.map((step) => ({
    step,
    views: events.filter(
      (e) => e.eventType === "step_view" && e.step === step,
    ).length,
  }));

  const inputFields = fields.filter(
    (f) => f.type !== "section_heading" && f.type !== "page_break",
  );

  const questions: QuestionAnalytics[] = inputFields.map((field) => {
    const values = responses.map((r) => r.answers[field.id]);
    const answeredCount = values.filter(
      (v) =>
        v !== null &&
        v !== undefined &&
        v !== "" &&
        !(Array.isArray(v) && v.length === 0),
    ).length;
    const emptyCount = responses.length - answeredCount;

    const q: QuestionAnalytics = {
      fieldId: field.id,
      type: field.type,
      label: field.label,
      answeredCount,
      emptyCount,
    };

    if (
      field.type === "single_choice" ||
      field.type === "multiple_choice" ||
      field.type === "dropdown"
    ) {
      const counts = new Map<string, number>();
      const options = field.config.options ?? [];
      for (const opt of options) counts.set(opt.value, 0);
      for (const v of values) {
        if (Array.isArray(v)) {
          for (const item of v) {
            counts.set(String(item), (counts.get(String(item)) ?? 0) + 1);
          }
        } else if (v !== null && v !== undefined && v !== "") {
          counts.set(String(v), (counts.get(String(v)) ?? 0) + 1);
        }
      }
      q.choiceDistribution = options.map((opt) => ({
        label: opt.label,
        count: counts.get(opt.value) ?? 0,
      }));
    }

    if (field.type === "rating") {
      const nums = values
        .map((v) => Number(v))
        .filter((n) => !isNaN(n) && n > 0);
      if (nums.length) {
        q.ratingAverage =
          Math.round((nums.reduce((a, b) => a + b, 0) / nums.length) * 10) /
          10;
      }
    }

    if (field.type === "nps") {
      const nums = values
        .map((v) => Number(v))
        .filter((n) => !isNaN(n) && n >= 0 && n <= 10);
      if (nums.length) {
        const promoters = nums.filter((n) => n >= 9).length;
        const passives = nums.filter((n) => n >= 7 && n <= 8).length;
        const detractors = nums.filter((n) => n <= 6).length;
        const score = Math.round(
          ((promoters - detractors) / nums.length) * 100,
        );
        q.npsScore = score;
        q.npsBreakdown = { promoters, passives, detractors };
      }
    }

    if (field.type === "short_text" || field.type === "long_text") {
      const recent: string[] = [];
      const wordFreq = new Map<string, number>();
      for (const v of values) {
        if (typeof v !== "string" || !v.trim()) continue;
        if (recent.length < 5) recent.push(v.trim());
        const words = v
          .toLowerCase()
          .replace(/[^\p{L}\p{N}\s]/gu, " ")
          .split(/\s+/)
          .filter(
            (w) =>
              w.length > 3 &&
              !STOP_WORDS.has(w),
          );
        for (const w of words) {
          wordFreq.set(w, (wordFreq.get(w) ?? 0) + 1);
        }
      }
      q.recentAnswers = recent;
      q.topWords = Array.from(wordFreq.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, 8)
        .map(([word, count]) => ({ word, count }));
    }

    return q;
  });

  return {
    totalViews,
    totalStarts,
    totalSubmissions,
    completionRate,
    avgCompletionSeconds,
    responsesByDay,
    dropoffByStep,
    questions,
  };
}

const STOP_WORDS = new Set([
  "the", "and", "for", "are", "but", "not", "you", "all", "can", "had",
  "her", "was", "one", "our", "out", "day", "get", "has", "him", "his",
  "how", "man", "new", "now", "old", "see", "two", "way", "who", "boy",
  "did", "its", "let", "put", "say", "she", "too", "use", "with", "from",
  "this", "that", "have", "they", "your", "what", "would", "there", "their",
  "which", "about", "just", "into", "than", "more", "very", "some", "like",
  "when", "these", "want", "been", "were", "could", "after", "then",
]);

export interface FunnelStep {
  step: number;
  views: number;
  /** Percentage retention vs first step. 0–100. */
  retention: number;
  /** Drop-off vs previous step (positive = lost). 0–100. */
  dropFromPrev: number;
}

export interface FunnelSummary {
  steps: FunnelStep[];
  /** % of starts that reached submit. */
  completionRate: number;
  /** Step where most respondents dropped off, if any. */
  worstStep: number | null;
}

export function computeFunnel({
  events,
  totalSubmissions,
}: {
  fields: FormFieldDef[];
  events: EventInput[];
  totalSubmissions: number;
}): FunnelSummary {
  const stepCounts = new Map<number, number>();
  for (const e of events) {
    if (e.eventType !== "step_view" || e.step == null) continue;
    stepCounts.set(e.step, (stepCounts.get(e.step) ?? 0) + 1);
  }

  const steps = Array.from(stepCounts.keys()).sort((a, b) => a - b);
  const first = steps[0] ? stepCounts.get(steps[0]) ?? 0 : 0;

  const out: FunnelStep[] = steps.map((step, i) => {
    const views = stepCounts.get(step) ?? 0;
    const prev = i > 0 ? stepCounts.get(steps[i - 1]) ?? 0 : views;
    return {
      step,
      views,
      retention: first > 0 ? Math.round((views / first) * 100) : 0,
      dropFromPrev:
        prev > 0 ? Math.max(0, Math.round(((prev - views) / prev) * 100)) : 0,
    };
  });

  const completionRate =
    first > 0 ? Math.round((totalSubmissions / first) * 100) : 0;

  let worstStep: number | null = null;
  let worstDrop = 0;
  for (const s of out) {
    if (s.dropFromPrev > worstDrop) {
      worstDrop = s.dropFromPrev;
      worstStep = s.step;
    }
  }

  return { steps: out, completionRate, worstStep };
}

export interface InteractionInsights {
  mostInteractedField: { fieldId: string; label: string; count: number } | null;
  mostErroredField: { fieldId: string; label: string; count: number } | null;
  worstStep: { step: number; dropOffs: number } | null;
  avgTimePerStepSeconds: { step: number; avgSeconds: number }[];
}

export function computeInteractionInsights({
  fields,
  events,
}: {
  fields: FormFieldDef[];
  events: EventInput[];
}): InteractionInsights {
  const byField = new Map<string, number>();
  const errorByField = new Map<string, number>();
  const dropOffs = new Map<number, number>();

  // Time per step: per-session, time between step_view events.
  const sessionStepTimes = new Map<string, Map<number, number>>(); // session -> step -> ms-spent
  const lastByStepBySession = new Map<string, { step: number; ts: number }>();

  for (const e of events) {
    if (e.eventType === "field_focus" && e.fieldId) {
      byField.set(e.fieldId, (byField.get(e.fieldId) ?? 0) + 1);
    }
    if (e.eventType === "field_change" && e.fieldId) {
      byField.set(e.fieldId, (byField.get(e.fieldId) ?? 0) + 1);
    }
    if (e.eventType === "field_error" && e.fieldId) {
      errorByField.set(
        e.fieldId,
        (errorByField.get(e.fieldId) ?? 0) + 1,
      );
    }
    if (e.eventType === "drop_off" && e.step != null) {
      dropOffs.set(e.step, (dropOffs.get(e.step) ?? 0) + 1);
    }
    if (e.eventType === "step_view" && e.step != null && e.sessionId) {
      const last = lastByStepBySession.get(e.sessionId);
      if (last && last.step !== e.step) {
        const elapsed = Math.max(0, e.createdAt - last.ts);
        const sm =
          sessionStepTimes.get(e.sessionId) ??
          new Map<number, number>();
        sm.set(last.step, (sm.get(last.step) ?? 0) + elapsed);
        sessionStepTimes.set(e.sessionId, sm);
      }
      lastByStepBySession.set(e.sessionId, { step: e.step, ts: e.createdAt });
    }
  }

  const fieldLabel = (id: string) =>
    fields.find((f) => f.id === id)?.label ?? id;

  const top = (m: Map<string, number>) => {
    let best: [string, number] | null = null;
    for (const entry of m.entries()) {
      if (!best || entry[1] > best[1]) best = entry;
    }
    return best;
  };

  const topInteract = top(byField);
  const topError = top(errorByField);

  let worstStep: { step: number; dropOffs: number } | null = null;
  for (const [step, count] of dropOffs.entries()) {
    if (!worstStep || count > worstStep.dropOffs) {
      worstStep = { step, dropOffs: count };
    }
  }

  // Aggregate avg time per step
  const stepSums = new Map<number, { sum: number; n: number }>();
  for (const sm of sessionStepTimes.values()) {
    for (const [step, ms] of sm.entries()) {
      const cur = stepSums.get(step) ?? { sum: 0, n: 0 };
      cur.sum += ms;
      cur.n += 1;
      stepSums.set(step, cur);
    }
  }
  const avgTimePerStepSeconds = Array.from(stepSums.entries())
    .sort((a, b) => a[0] - b[0])
    .map(([step, v]) => ({
      step,
      avgSeconds: v.n > 0 ? Math.round(v.sum / v.n / 1000) : 0,
    }));

  return {
    mostInteractedField: topInteract
      ? {
          fieldId: topInteract[0],
          label: fieldLabel(topInteract[0]),
          count: topInteract[1],
        }
      : null,
    mostErroredField: topError
      ? {
          fieldId: topError[0],
          label: fieldLabel(topError[0]),
          count: topError[1],
        }
      : null,
    worstStep,
    avgTimePerStepSeconds,
  };
}

File: backend/app/services/analytics.py
from datetime import datetime, timedelta, timezone
from typing import Any

STOP_WORDS = {
    "the", "and", "for", "are", "but", "not", "you", "all", "can", "had",
    "her", "was", "one", "our", "out", "day", "get", "has", "him", "his",
    "how", "man", "new", "now", "old", "see", "two", "way", "who", "with",
    "from", "this", "that", "have", "they", "your", "what", "would", "there",
}


def compute_analytics(
    fields: list[dict[str, Any]],
    events: list[dict[str, Any]],
    responses: list[dict[str, Any]],
    range_days: int = 14,
) -> dict[str, Any]:
    total_views = sum(1 for e in events if e["eventType"] == "view")
    total_starts = sum(1 for e in events if e["eventType"] == "start")
    total_submissions = len(responses)

    if total_starts:
        completion_rate = min(100, round((total_submissions / total_starts) * 100))
    elif total_views:
        completion_rate = round((total_submissions / total_views) * 100)
    else:
        completion_rate = 0

    times = [
        r["completionTimeSeconds"]
        for r in responses
        if isinstance(r.get("completionTimeSeconds"), (int, float)) and r["completionTimeSeconds"] > 0
    ]
    avg_completion_seconds = round(sum(times) / len(times)) if times else 0

    today = datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0)
    responses_by_day = []
    for i in range(range_days - 1, -1, -1):
        day = today - timedelta(days=i)
        key = day.strftime("%Y-%m-%d")
        count = sum(
            1
            for r in responses
            if datetime.fromtimestamp(r["submittedAt"] / 1000, tz=timezone.utc).strftime("%Y-%m-%d") == key
        )
        responses_by_day.append({"date": key, "count": count})

    steps = sorted({f["step"] for f in fields if f.get("step", 0) > 0})
    dropoff_by_step = [
        {
            "step": step,
            "views": sum(1 for e in events if e["eventType"] == "step_view" and e.get("step") == step),
        }
        for step in steps
    ]

    input_fields = [f for f in fields if f["type"] not in ("section_heading", "page_break")]
    questions = [_question_stats(f, responses) for f in input_fields]

    return {
        "totalViews": total_views,
        "totalStarts": total_starts,
        "totalSubmissions": total_submissions,
        "completionRate": completion_rate,
        "avgCompletionSeconds": avg_completion_seconds,
        "responsesByDay": responses_by_day,
        "dropoffByStep": dropoff_by_step,
        "questions": questions,
    }


def _question_stats(field: dict[str, Any], responses: list[dict[str, Any]]) -> dict[str, Any]:
    values = [r["answers"].get(field["id"]) for r in responses]
    answered = sum(
        1
        for v in values
        if v is not None and v != "" and not (isinstance(v, list) and len(v) == 0)
    )
    q: dict[str, Any] = {
        "fieldId": field["id"],
        "type": field["type"],
        "label": field["label"],
        "answeredCount": answered,
        "emptyCount": len(responses) - answered,
    }
    ftype = field["type"]
    if ftype in ("single_choice", "multiple_choice", "dropdown"):
        options = (field.get("config") or {}).get("options") or []
        counts = {opt["value"]: 0 for opt in options}
        for v in values:
            if isinstance(v, list):
                for item in v:
                    counts[str(item)] = counts.get(str(item), 0) + 1
            elif v not in (None, "", []):
                counts[str(v)] = counts.get(str(v), 0) + 1
        q["choiceDistribution"] = [
            {"label": opt["label"], "count": counts.get(opt["value"], 0)} for opt in options
        ]
    if ftype == "rating":
        nums = [float(v) for v in values if _is_num(v) and float(v) > 0]
        if nums:
            q["ratingAverage"] = round(sum(nums) / len(nums), 1)
    if ftype == "nps":
        nums = [int(v) for v in values if _is_num(v) and 0 <= int(v) <= 10]
        if nums:
            promoters = sum(1 for n in nums if n >= 9)
            detractors = sum(1 for n in nums if n <= 6)
            passives = len(nums) - promoters - detractors
            q["npsScore"] = round(((promoters - detractors) / len(nums)) * 100)
            q["npsBreakdown"] = {"promoters": promoters, "passives": passives, "detractors": detractors}
    if ftype in ("short_text", "long_text"):
        recent, word_freq = [], {}
        for v in values:
            if not isinstance(v, str) or not v.strip():
                continue
            if len(recent) < 5:
                recent.append(v.strip())
            for w in _words(v):
                if len(w) > 3 and w not in STOP_WORDS:
                    word_freq[w] = word_freq.get(w, 0) + 1
        q["recentAnswers"] = recent
        q["topWords"] = sorted(word_freq.items(), key=lambda x: -x[1])[:8]
        q["topWords"] = [{"word": w, "count": c} for w, c in q["topWords"]]
    return q


def _is_num(v: Any) -> bool:
    try:
        float(v)
        return True
    except (TypeError, ValueError):
        return False


def _words(text: str) -> list[str]:
    import re
    return re.sub(r"[^\w\s]", " ", text.lower()).split()


def compute_funnel(events: list[dict[str, Any]], total_submissions: int) -> dict[str, Any]:
    step_counts: dict[int, int] = {}
    for e in events:
        if e["eventType"] == "step_view" and e.get("step") is not None:
            step_counts[e["step"]] = step_counts.get(e["step"], 0) + 1
    steps_sorted = sorted(step_counts.keys())
    first = step_counts.get(steps_sorted[0], 0) if steps_sorted else 0
    out = []
    for i, step in enumerate(steps_sorted):
        views = step_counts[step]
        prev = step_counts.get(steps_sorted[i - 1], views) if i > 0 else views
        out.append({
            "step": step,
            "views": views,
            "retention": round((views / first) * 100) if first else 0,
            "dropFromPrev": max(0, round(((prev - views) / prev) * 100)) if prev else 0,
        })
    completion_rate = round((total_submissions / first) * 100) if first else 0
    worst = max(out, key=lambda s: s["dropFromPrev"], default=None)
    return {
        "steps": out,
        "completionRate": completion_rate,
        "worstStep": worst["step"] if worst and worst["dropFromPrev"] > 0 else None,
    }


def compute_interaction_insights(
    fields: list[dict[str, Any]], events: list[dict[str, Any]]
) -> dict[str, Any]:
    by_field: dict[str, int] = {}
    error_by_field: dict[str, int] = {}
    drop_offs: dict[int, int] = {}
    session_step_times: dict[str, dict[int, float]] = {}
    last_by_session: dict[str, dict[str, Any]] = {}

    for e in events:
        et = e["eventType"]
        fid = e.get("fieldId")
        if et in ("field_focus", "field_change") and fid:
            by_field[fid] = by_field.get(fid, 0) + 1
        if et == "field_error" and fid:
            error_by_field[fid] = error_by_field.get(fid, 0) + 1
        if et == "drop_off" and e.get("step") is not None:
            drop_offs[e["step"]] = drop_offs.get(e["step"], 0) + 1
        if et == "step_view" and e.get("step") is not None and e.get("sessionId"):
            sid = e["sessionId"]
            last = last_by_session.get(sid)
            if last and last["step"] != e["step"]:
                elapsed = max(0, e["createdAt"] - last["ts"])
                session_step_times.setdefault(sid, {})
                session_step_times[sid][last["step"]] = session_step_times[sid].get(last["step"], 0) + elapsed
            last_by_session[sid] = {"step": e["step"], "ts": e["createdAt"]}

    labels = {f["id"]: f["label"] for f in fields}

    def top(m: dict[str, int]) -> tuple[str, int] | None:
        if not m:
            return None
        return max(m.items(), key=lambda x: x[1])

    ti = top(by_field)
    te = top(error_by_field)
    worst_step = max(drop_offs.items(), key=lambda x: x[1], default=None)

    step_sums: dict[int, dict[str, float]] = {}
    for sm in session_step_times.values():
        for step, ms in sm.items():
            step_sums.setdefault(step, {"sum": 0, "n": 0})
            step_sums[step]["sum"] += ms
            step_sums[step]["n"] += 1

    return {
        "mostInteractedField": (
            {"fieldId": ti[0], "label": labels.get(ti[0], ti[0]), "count": ti[1]} if ti else None
        ),
        "mostErroredField": (
            {"fieldId": te[0], "label": labels.get(te[0], te[0]), "count": te[1]} if te else None
        ),
        "worstStep": {"step": worst_step[0], "dropOffs": worst_step[1]} if worst_step else None,
        "avgTimePerStepSeconds": [
            {"step": s, "avgSeconds": round(v["sum"] / v["n"] / 1000) if v["n"] else 0}
            for s, v in sorted(step_sums.items())
        ],
    }
